s3transfer/utils.py

```diff
@@ -499,13 +500,29 @@ def enable_callback(self):
     def disable_callback(self):
         self._callbacks_enabled = False
 
-    def seek(self, where):
-        self._fileobj.seek(self._start_byte + where)
+    def seek(self, where, whence=0):
+        if whence not in (0, 1, 2):
+            # Mimic io's error for invalid whence values
+            raise ValueError(
+                "invalid whence (%s, should be 0, 1 or 2)" % whence)
+
+        # Recalculate where based on chunk attributes so seek from file
+        # start (whence=0) is always used
+        where += self._start_byte
+        if whence == 1:
+            where += self._amount_read
+        elif whence == 2:
+            where += self._size
+
+        self._fileobj.seek(max(where, 0 + self._start_byte))
         if self._callbacks is not None and self._callbacks_enabled:
             # To also rewind the callback() for an accurate progress report
+            bounded_where = max(min(where - self._start_byte, self._size), 0)
+            bounded_amount_read = min(self._amount_read, self._size)
+            amount = bounded_where - bounded_amount_read
             invoke_progress_callbacks(
-                self._callbacks, bytes_transferred=where - self._amount_read)
-        self._amount_read = where
+                self._callbacks, bytes_transferred=amount)
+        self._amount_read = max(where - self._start_byte, 0)
```

Comment:
It is a little weird that the value `_amount_read` can be outside of the read chunk boundaries, but I don't think we need to rename the variable for the sake of minimizing the amount of changes we are making here. Instead, I think we should add a comment here saying that `_amount_read` is more synonymous with a tell position and can be outside of the boundary as opposed to the actual amount of data read.